iwyu.cc

```diff
@@ -2559,23 +2559,19 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     // declarations are forward-declarable.  This is true, *except*
     // for the exception (throw) types.  We clean that up here.
     // TODO(csilvers): figure out how to do these two steps in one place.
+    ASTNode* ast_node = current_ast_node();
     const FunctionProtoType* fn_type = nullptr;
-    if (!fn_type) {
-      fn_type = current_ast_node()->template GetParentAs<FunctionProtoType>();
-    }
-    if (!fn_type) {
-      if (const FunctionDecl* fn_decl
-          = current_ast_node()->template GetParentAs<FunctionDecl>())
-        fn_type = dyn_cast<FunctionProtoType>(GetTypeOf(fn_decl));
+    if (const auto* fn_decl = ast_node->GetAncestorOfType<FunctionDecl>()) {
+      fn_type = dyn_cast<FunctionProtoType>(GetTypeOf(fn_decl));
     }
+
```

Comment:
It has been a while and not recall exactly, but could it be that there is only a `FunctionProtoType` without a corresponding `FunctionDecl`? Because in that case the new code won't find the prototype.